### ISSUE TITLE
CLI attempts NuGet Restore instead of using provided DLL path

### ISSUE THREAD

**Issue by mwpowellhtx** [dotnet/cli#11687]
## Background

I have not published my tooling project to NuGet yet. I plan to today, but I have a curious message, error, or warning, that is happening at the moment.

I uniquely renamed my project output such that it no longer conflicts with existing NuGet packages. I am finding this kind of error during the build. I think that the local path is being used, but the error in the meantime is curious. I'm not sure there is anything else I can do with that from my end, when the build resolves, the asset is *there*, so to speak.

```
Error / NU1101 / Unable to find package dotnet-mytool. No packages exist with this id in source(s): ... / MyTool.Integration.Tests / path\to\MyTool.Integration.Tests\MyTool.Integration.Tests.csproj
```

That being said on the one hand, I'm not sure my targets do not need a bit of fine tuning. There is one in particular depending on some curious *Microsoft Build* targets, which are triggered at odd times when the asset might not be *there*, i.e. preparatory phases when the build outputs might not otherwise be there, for instance.

So, really, I'm just curious, if we provide a full path to the *DLL* assets, isn't that sufficient to preclude this sort of condition? The default position, then, would be to ask for a *NuGet Restore*, for example.

## Environment data
Attached.

[dotnet-info.txt](https://github.com/dotnet/cli/files/3346921/dotnet-info.txt)

**Comment by livarcocc** [comment]
Hey @mwpowellhtx, I am sorry, I don't quite follow exactly what you are trying to do.

Is this a project tool, a global tool or a local tool? What does your csproj look like (if this is a project tool)?

How are you invoking it exactly? Where would you provide the DLL assets that you mention above?

**Comment by mwpowellhtx** [comment]
@livarcocc Again, not sure what you mean by project v local v global tools. During development it is intended that I test against the project output. Eventually, I intend to publish, at which point, for subscribers, I suppose it would be a NuGet-delivered tool through whatever NuGet channels are appropriate, i.e. solution ``package/`` directory.

**Comment by livarcocc** [comment]
Project tools are specified as <DotnetCLiToolReference> in the project file. They can be invoked by running dotnet toolname when in the project folder.

Global tools are shared framework apps packaged in a special way. They are installed through dotnet tool install -g and can be invoked through toolname.

Local tools are the same package as global tools but they are installed to a manifest file somewhere in your repo and can be invoked through dotnet toolname as well.

What command are you running that is giving you the error message above. Any change you can give us a repro repo on github that we can clone and run some command(s) to replicate your issue?

**Comment by mwpowellhtx** [comment]
@livarcocc In this case, as an end-to-end integration test, I am literally trying to invoke:

```
dotnet path\to\dotnet-mytool.dll ...
                                 ^ arguments
```

After the build has completed, no problem, the path-oriented asset receives the correct instruction, and my integration tests run fine. Leading up to that point, the asset may not have been built yet, so I think ``dotnet`` is attempting to fall back on a package from NuGet, which in this case, would not exist yet. This is incorrect, I do not want dotnet to do anything further if it cannot identify the path exists.

In other words, there is a reason I have identified the path to the project output. This is an end-to-end integration test.

Make sense?

**Comment by peterhuene** [comment]
Try using `dotnet exec <path_to_dll> ...`.

**Comment by mwpowellhtx** [comment]
@peterhuene No effect, same outcome. However, it does not seem to be a critical error, and is understandable given my build pipeline.

**Comment by mwpowellhtx** [comment]
Which, to be clear, I do not read anything about ``dotnet exec`` failing anything in the command line help. I just think that may be the default ``dotnet`` *SDK command* behavior. Oddly enough, ``exec`` not listed among those commands.

**Comment by peterhuene** [comment]
> No effect, same outcome. 

`dotnet exec <dll>` shouldn't have the same effect; it should always fail with `The application to execute does not exist` if the path to the application dll does not exist.  This differs from just `dotnet <dll>` because if the application dll doesn't exist, it passes the path to the .NET Core SDK to interpret as a command, which may be a built-in command, a `dotnet-` prefixed executable on the `PATH`, or a [project tool](https://docs.microsoft.com/en-us/dotnet/core/tools/extensibility#per-project-based-extensibility).  The end result is likely to be the same (i.e. that the command failed to execute), but the .NET Core SDK does a number of additional steps to resolve the argument as an internally or externally valid command.

As `dotnet exec` is built-in to `dotnet` itself and not part of the .NET Core SDK itself, it doesn't show up in `dotnet help`, which is the help output of the .NET Core SDK and not `dotnet`.  It's a bit confusing, I can understand.

At any rate, let me step back and try to better understand the problem you are facing.  If you're looking for dotnet to simply fail if the application dll doesn't exist, then `dotnet exec` is your best option.

**Comment by mwpowellhtx** [comment]
@peterhuene Okay, well, I tried with and without the switches, same outcome. Reported via my build log:

```
Evaluating `dotnet  exec ..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll --version´ ...
```

With the targets commands:

```Xml
<Message Text="Evaluating `dotnet $(DotnetMyToolPath) --version´ ..." Importance="high" />
<Exec Command="dotnet $(DotnetMyToolPath) --version" ConsoleToMSBuild="true" StandardOutputImportance="normal" ContinueOnError="true">
  <Output TaskParameter="ConsoleOutput" PropertyName="_MyToolVersion" />
</Exec>
```

``DotnetMyToolPath`` defined as : ``exec path\to\dotnet-mytool.dll``, with the error:

```
Error / NU1101 / Unable to find package dotnet-mytool. No packages exist with this id in source(s): ...
```

So, I hear you, *it shouldn't*, however, the evidence is clear, ***IT DOES***. That is, attempt to peer into package sources.

The error does not appear to be fatal to ``dotnet``, and I think at this point this is just a natural response to the my build pipeline as I have it arranged. It is probably fine, I would just like to clean up some of these sort of false positives and/or false negatives.

**Comment by mwpowellhtx** [comment]
@peterhuene What are you basing your assumptions on?

**Comment by peterhuene** [comment]
Hi @mwpowellhtx.  Do you get the error if you manually run `dotnet exec path\to\dotnet-mytool.dll --version` outside of the context of MSBuild?

If this project has a `DotNetCliToolReference` item and any restoring command (i.e. `dotnet restore`, `dotnet build`, `dotnet publish`, etc.) is run without the tool being available in a NuGet source, I would expect a `NU1101` error from the restore operation; however, it wouldn't explain why your target containing your `Exec` task is executing since the restore itself should be failing.

What I would not expect is that this error is coming from `dotnet exec`.  That command does not invoke a restore operation or MSBuild, so it can't generate a `NU1101`.  If the path specified to `dotnet exec` is not available, no managed code will execute at all.

If you like, I'd be happy to look at a [MSBuild binlog](https://gist.github.com/dsplaisted/b256d7804672a6f6375476a5f5658b7b) for you via email (my address is pehuene@microsoft.com).  Please be aware, though, that an MSBuild binlog will collect information such as environment variables, file paths, and other build-related information.

**Comment by mwpowellhtx** [comment]
@peterhuene Not at the command line. As mentioned here, or in an earlier issue, the error is obviously not of ``dotnet`` origin, so I suspect this probably has to do with one surrogate *Microsoft Build* target or another, attempting to invoke a *NuGet* target.

```
G:\Source\Spikes\MyTooling\Prototype\src\MyTool\bin\Release\netcoreapp2.1>dotnet exec .\dotnet-mytool.dll --version
1.0.0.24806
```

Whereas, when I do a similar thing, I do in fact get an error:

```
G:\Source\Spikes\MyTooling\Prototype\src\MyTool\bin\Release\netcoreapp2.1>dotnet exec .\dotnet-testing.dll --version
The application to execute does not exist: '.\dotnet-testing.dll'
```

**Comment by peterhuene** [comment]
@mwpowellhtx were your able to resolve your issue? I just want to make sure nothing is blocking your scenario.

If you would still like me to inspect your build log, feel free to email one to me at any time. Thanks for taking the time to report your issue to us!

**Comment by mwpowellhtx** [comment]
@peterhuene Sure no problem. Honestly, I think it is probably normal behavior, not fatal to the surrogate NuGet invocation, builds succeed anyway. Would be better if the *error* could be simply ignored, not reported, etc, but whatever.

**Comment by mwpowellhtx** [comment]
@peterhuene I do not know how much of a corner case this is. Of course, now that I have published, *MSBuild* pulls those packages down, but would not use them because I furnish the path internal to the solution, but still, can be confusing.

**Comment by mwpowellhtx** [comment]
@peterhuene I have two issues, the first phase described here. The second is the prototypical ``Package XYZ is not compatible with netcoreapp2.1``, related to the same. I do not think it is a CLI issue, per se, but I do think it is *MSBuild*, *dotnet*, or one of its surrogates, i.e. *NuGet*, related. I can be a little less cryptic now that I have deployed if you would like to [please take a look](/mwpowellhtx/Code.Generation.Roslyn) and let me know what you think. I have partially traced some of the problem to a problem during packaging of the assemblies themselves, is literally using the ``PackageVersion`` instead of the version range I have specified, which should be something like this: *[2.10.0,3.0.0)*. Packages in question are: *System.Reflection.Metadata 1.0.21*, *System.Collections.Immutable 1.1.36*, *Microsoft.CodeAnalysis.CSharp 1.0.1*, *Microsoft.CodeAnalysis.Common 1.0.1*. You can discover this by unzipping the actual package and inspecting the ``nuspec`` that *dotnet*/*nuget* thought was there. But like I said, if the version was correct, then yes, these are likely legit errors; but the dependency version is incorrect. Otherwise, other than that packaging snafu, actually subscribing to them in a real world scenario does not seem to be a problem.

**Comment by mwpowellhtx** [comment]
Possibly is a second issue, above. Possibly not a CLI issue, per se, more likely an *MSBuild* issue, *dotnet*, or *nuget*.

**Comment by mwpowellhtx** [comment]
@peterhuene May have just identified an actual typo in my dependencies.
